### ISSUE TITLE
v1.23.5 regression: `ddev composer` command error when passing JSON or arg with spaces

### ISSUE THREAD

**Issue by stasadev** [ddev/ddev#6628]
### Preliminary checklist

- [x] I am using the latest stable version of DDEV (see [upgrade guide](https://ddev.readthedocs.io/en/stable/users/install/ddev-upgrade/))
- [x] I have searched [existing issues](https://github.com/ddev/ddev/issues)
- [x] I have checked the [troubleshooting guide](https://ddev.readthedocs.io/en/stable/users/usage/troubleshooting/)
- [x] I have run `ddev debug test` to include output below

### Output of `ddev debug test`

<details><summary>Expand `ddev debug test` diagnostic information</summary>

```
[COPY-PASTE HERE THE OUTPUT OF `ddev debug test`]
```
</details>

### Expected Behavior

`ddev composer` should work as `composer` inside the container.

### Actual Behavior

Since today (1.23.5), it fails with this message: 

```
In ConfigCommand.php line 766:You must pass the type and a url. Example: php composer.phar config reposit  ories.foo vcs https://bar.com
```
As I'm passing a type and url, I suspect there are lost somewhere.

### Steps To Reproduce

From Discord https://discord.com/channels/664580571770388500/1296719909367054420

```
ddev composer config repositories.0 "{\"type\": \"composer\", \"url\":\"hello\",  \"exclude\": [\"magento/composer-dependency-version-audit-plugin\"]}"
```

(We can pass a json)

### Anything else?

I've just found a workaround: 

```
ddev exec --raw composer  config repositories.0 "{\"type\": \"composer\", \"url\":\"hello\",  \"exclude\": [\"magento/composer-dependency-version-audit-plugin\"]}"
```

**Comment by stasadev** [comment]
The fix is available in:

- https://github.com/ddev/ddev/pull/6627#issuecomment-2422346715

**Comment by rfay** [comment]
Aw, we did look at that with a little bit of worry :(

**Comment by rfay** [comment]
I don't understand:
* Why I took the risk of changing that
* Why I *needed* to take the risk of changing from raw (we should look together)
* How this class of problem can haunt us for years and years :(

**Comment by stasadev** [comment]
> Why I took the risk of changing that
> Why I *needed* to take the risk of changing from raw (we should look together)

Because only `Cmd` uses `$PATH` from `bash`:

https://github.com/ddev/ddev/blob/a6d7d237c0a403e5d878b8378bf24f3d7f5ab8eb/pkg/ddevapp/ddevapp.go#L2050-L2059

And we add `$PATH` only for bash, not anything else

https://github.com/rfay/ddev/blob/3a6e225559088992d659f343860485da8fab9c94/containers/ddev-webserver/ddev-webserver-base-files/etc/bash.bashrc#L9

> How this class of problem can haunt us for years and years :( 

I didn't see a problem there because we didn't have test coverage.

---

The question I ask myself: how can we add more items to `$PATH` for every shell, not only for bash?
If we can resolve it, then we can switch back to `RawCmd` without workarounds.

I update `$PATH` on Arch Linux in `/etc/profile.d`, and I will look if this can be applied to Debian.

**Comment by stasadev** [comment]
> I update `$PATH` on Arch Linux in `/etc/profile.d`, and I will look if this can be applied to Debian.

It didn't work, because `$PATH` is not updated in `docker-compose exec`.

ChatGPT answer:

When using `docker-compose exec`, it launches a non-login, non-interactive shell by default. This prevents it from sourcing environment configuration files like `/etc/profile` or scripts in `/etc/profile.d` (which are typically sourced by login shells).

So instead we need to pass `$PATH` that we have to `docker-compose exec`, I'll update the PR.

**Comment by stasadev** [comment]
I used a `.ddev/web-build/Dockerfile` to understand it better:

```dockerfile
RUN echo $PATH
```
I see this in `ddev debug rebuild`:
```
#9 [web 6/7] RUN echo /usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
#9 0.299 ~/bin:/var/www/html/vendor/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/var/www/html/bin
```
It has two different values at different moments of execution 🤯.

And overriding `/etc/profile` didn't work, I tried to build a new image from `debian:bookworm-slim` with different `/etc/profile` and then:

```
$ docker exec -it debian-slim env | grep PATH
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

I even looked at the official image source to check if they have `ENV PATH=...` somewhere
https://github.com/debuerreotype/docker-debian-artifacts/tree/dist-amd64/bookworm/slim
But there is nothing.

Then I checked `busybox`, `alpine` and saw the same `$PATH` everywhere.

Finally I looked in `moby/moby`, and they have a lot of `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin`

**It's their default**:

https://github.com/moby/moby/blob/7f8ef5c09df9f24d869fc7da1f0fef8fb3047681/vendor/github.com/moby/buildkit/util/system/path.go#L11-L14

So `docker-compose exec` use the same `$PATH` everywhere, and we can only change it in two cases:
1. Override the env, just like I did in the related PR, `docker-compose exec -e PATH=... composer ...`
2. Run `docker-compose exec bash -c "..."` and bash will read the PATH from the container files.

I think we're doing 100% of what we can here.

**Comment by rfay** [comment]
So the bottom line is that `docker-compose exec` is the culprit, not using any of the normal Linux techniques. Wow. I wonder if we should raise an issue with docker compose?

**Comment by stasadev** [comment]
I also see the same behavior with `docker exec`.

I don't know if this is a bug or expected behavior, I can only say that Docker has several default variables for `/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin` and this value is used a lot in their tests.

**Comment by rfay** [comment]
Sounds like it's probably `docker exec` then. But I would sure expect `docker exec` to behave like non-interactive non-login. We also might be able to change our approach to behave more like `ddev ssh` does,
```
		err = app.ExecWithTty(&ddevapp.ExecOpts{
			Service: serviceType,
			Cmd:     shell + " -l",
			Dir:     sshDirArg,
		})
```

**Comment by HeikkiYlipaavalniemi** [comment]
I encountered similar issues after updating to latest version of DDEV. In my case I tested two different projects and in both projects just the `ddev composer install` command was failing. 

The workarond from above using `ddev exec --raw composer install` worked for me also. Looking at your discussion it seems you are on top of the problem but I wanted to comment because the issue title seems to indicate that the problem would only be in certain Composer commands but in my case even the normal install command is failing. I tested the install command several times and also made sure I powered off and restarted DDEV. Only way I could get the install command to pass was with the exec command.

**Comment by stasadev** [comment]
> I encountered similar issues after updating to latest version of DDEV. In my case I tested two different projects and in both projects just the `ddev composer install` command was failing.

Hi @HeikkiYlipaavalniemi,

That's unexpected. But I can't say more without looking at `composer.json`.

> Looking at your discussion it seems you are on top of the problem

Yes, we will revert to the original behavior in the next DDEV release, sorry for the inconvenience.
In the meantime you can use the updated fix from https://github.com/ddev/ddev/pull/6627#issuecomment-2422346715

**Comment by HeikkiYlipaavalniemi** [comment]
I debugged this further and on further investigation I noticed that some of the rows in the composer.json had white space issues where the JSON wasn't indented correctly. After I fixed the white space issues in the JSON it seemed like the command started to work again.

Perhaps the spaces are causing the issue also if the actual JSON file has wrong spacing inside it?

Weirdly after I fixed the white spaces and then changed branch it was still able to run the command successfully.

But at least for now it seems the problem is solved for us with the current version also.
